pkg/config/msp.go

```diff
@@ -0,0 +1,264 @@
+/*
+Copyright IBM Corp. All Rights Reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package config
+
+import (
+	"crypto/ecdsa"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/asn1"
+	"encoding/pem"
+	"fmt"
+
+	mb "github.com/hyperledger/fabric-protos-go/msp"
+)
+
+// MSP is the configuration information for
+// a Fabric MSP.
+// Here we assume a default certificate validation policy, where
+// any certificate signed by any of the listed rootCA certs would
+// be considered as valid under this MSP.
+// This MSP may or may not come with a signing identity. If it does,
+// it can also issue signing identities. If it does not, it can only
+// be used to validate and verify certificates.
+type MSP struct {
+	// Name holds the identifier of the MSP; MSP identifier
+	// is chosen by the application that governs this MSP.
+	// For example, and assuming the default implementation of MSP,
+	// that is X.509-based and considers a single Issuer,
+	// this can refer to the Subject OU field or the Issuer OU field.
+	Name string
+	// List of root certificates trusted by this MSP
+	// they are used upon certificate validation (see
+	// comment for IntermediateCerts below).
+	RootCerts []x509.Certificate
+	// List of intermediate certificates trusted by this MSP;
+	// they are used upon certificate validation as follows:
+	// validation attempts to build a path from the certificate
+	// to be validated (which is at one end of the path) and
+	// one of the certs in the RootCerts field (which is at
+	// the other end of the path). If the path is longer than
+	// 2, certificates in the middle are searched within the
+	// IntermediateCerts pool.
+	IntermediateCerts []x509.Certificate
+	// Identity denoting the administrator of this MSP.
+	Admins []x509.Certificate
+	// Identity revocation list.
+	RevocationList []pkix.CertificateList
+	// SigningIdentity holds information on the signing identity
+	// this peer is to use, and which is to be imported by the
+	// MSP defined before.
+	SigningIdentity *SigningIdentityInfo
+	// OrganizationalUnitIdentifiers holds one or more
+	// fabric organizational unit identifiers that belong to
+	// this MSP configuration.
+	OrganizationalUnitIdentifiers []*OUIdentifier
+	// CryptoConfig contains the configuration parameters
+	// for the cryptographic algorithms used by this MSP.
+	CryptoConfig *CryptoConfig
+	// List of TLS root certificates trusted by this MSP.
+	// They are returned by GetTLSRootCerts.
+	TLSRootCerts []x509.Certificate
+	// List of TLS intermediate certificates trusted by this MSP;
+	// They are returned by GetTLSIntermediateCerts.
+	TLSIntermediateCerts []x509.Certificate
+	// fabric_node_ous contains the configuration to distinguish clients from peers from orderers
+	// based on the OUs.
+	NodeOus *NodeOUs
```

Comment:
values or references? In the case of the slices, they are references but should they contain pointers or values?